app/src/main/java/org/thoughtcrime/securesms/providers/PartProvider.java

```
@@ -0,0 +1,170 @@
+import org.thoughtcrime.securesms.database.DatabaseFactory;
+import org.thoughtcrime.securesms.mms.PartUriParser;
+import org.thoughtcrime.securesms.service.KeyCachingService;
+import org.thoughtcrime.securesms.util.MemoryFileUtil;
+import org.thoughtcrime.securesms.util.Util;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public final class PartProvider extends BaseContentProvider {
+
+  private static final String TAG = Log.tag(PartProvider.class);
+
+  private static final String CONTENT_AUTHORITY  = BuildConfig.APPLICATION_ID + ".part";
+  private static final String CONTENT_URI_STRING = "content://" + CONTENT_AUTHORITY + "/part";
+  private static final Uri    CONTENT_URI        = Uri.parse(CONTENT_URI_STRING);
+  private static final int    SINGLE_ROW         = 1;
+
+  private static final UriMatcher uriMatcher;
+
+  static {
+    uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);
+    uriMatcher.addURI(CONTENT_AUTHORITY, "part/*/#", SINGLE_ROW);
+  }
+
+  @Override
+  public boolean onCreate() {
+    Log.i(TAG, "onCreate()");
+    return true;
+  }
+
+  public static Uri getContentUri(AttachmentId attachmentId) {
+    Uri uri = Uri.withAppendedPath(CONTENT_URI, String.valueOf(attachmentId.getUniqueId()));
+    return ContentUris.withAppendedId(uri, attachmentId.getRowId());
+  }
+
+  @Override
+  public ParcelFileDescriptor openFile(@NonNull Uri uri, @NonNull String mode) throws FileNotFoundException {
+    Log.i(TAG, "openFile() called!");
+
+    if (KeyCachingService.isLocked(getContext())) {
+      Log.w(TAG, "masterSecret was null, abandoning.");
+      return null;
+    }
+
+    if (uriMatcher.match(uri) == SINGLE_ROW) {
+      Log.i(TAG, "Parting out a single row...");
+      try {
+        final PartUriParser partUri = new PartUriParser(uri);
+        return getParcelStreamForAttachment(partUri.getPartId());
+      } catch (IOException ioe) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)